homeassistant/components/zwave_js/addon.py

```diff
@@ -0,0 +1,256 @@
+"""Provide add-on management."""
+from __future__ import annotations
+
+import asyncio
+from functools import partial
+from typing import Callable, Optional, cast
+
+from awesomeversion import AwesomeVersion
+
+from homeassistant.components.hassio import (
+    async_get_addon_discovery_info,
+    async_get_addon_info,
+    async_install_addon,
+    async_set_addon_options,
+    async_start_addon,
+    async_stop_addon,
+    async_uninstall_addon,
+    async_update_addon,
+)
+from homeassistant.components.hassio.handler import HassioAPIError
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady, HomeAssistantError
+from homeassistant.helpers.singleton import singleton
+
+from .const import (
+    ADDON_SLUG,
+    CONF_ADDON_DEVICE,
+    CONF_ADDON_NETWORK_KEY,
+    CONF_NETWORK_KEY,
+    CONF_USB_PATH,
+    DOMAIN,
+    LOGGER,
+)
+
+DATA_ADDON_MANAGER = f"{DOMAIN}_addon_manager"
+
+
+@singleton(DATA_ADDON_MANAGER)
+@callback
+def get_addon_manager(hass: HomeAssistant) -> AddonManager:
+    """Get the add-on manager."""
+    return AddonManager(hass)
+
+
+async def async_ensure_addon_running(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Ensure that Z-Wave JS add-on is installed and running."""
+    addon_manager: AddonManager = get_addon_manager(hass)
+    try:
+        addon_is_installed = await addon_manager.async_is_addon_installed()
+        addon_is_running = await addon_manager.async_is_addon_running()
```

Comment:
This one should check if an update is in progress, because raising `ConfigEntryNotReady` will have HA retry the setup before the setup is done. If it's updating, raise too.

Reply:
> Also, no need to check `is_running` if you're already going to raise for `addon_is_installed`.

If `async_is_addon_installed` raised `async_is_addon_running` won't be run.

Reply:
I've added a check for add-on tasks in progress in the integration add-on helper functions.

Reply:
What I meant with my comment is that you're calling `await addon_manager.async_is_addon_running()` while you might already abort based on the result of `addon_is_installed`